src/flows/pipes/Visualization/threshold.ts

```diff
@@ -4,85 +4,141 @@ export enum ThresholdFormat {
   Deadman = 'deadman',
 }
 
+export enum ThresholdType {
+  Greater = 'greater',
+  GreaterEqual = 'greater-equal',
+  Less = 'less',
+  LessEqual = 'less-equal',
+  Equal = 'equal',
+  NotEqual = 'not-equal',
+  Between = 'between',
+  NotBetween = 'not-between',
+  Deadman = 'missing-for-longer-than',
+}
+
 export type Threshold = {
+  // value, min, and max are strings
+  // can be string numbers (e.g. '123.2') or lexical strings (e.g. 'a' < 'b')
   value: number
-  type: string
+  type: ThresholdType
   field: string
   max?: number
   min?: number
   deadmanCheckValue?: string
-  deadmanStopValue: string
+  deadmanStopValue: string // e.g. '5 minutes'
 }
 
 export interface ErrorThreshold extends Threshold {
   fieldType: string
 }
 
+const existenceCheck = unknown => unknown == 0 || !!unknown
+
+export function validateThreshold(t: Threshold): boolean {
```

Comment:
if all types have a validation, why isn't validation a part of the type specification?

Reply:
Additional schema-based validation will be a part of the next PR.
Which will confirm:
* what the monitor.check() will permit in the lambda. (a.k.a. flux accepts string equivalency...but want to check for specifics with the check method)
* grabbing the implicit schema. probably similar to how MetricsSelector does it.

Reply:
unless what you are asking for, is something else?

Reply:
oh definitely something else. so we had this problem we first identified within our redux reducers where switches grow... to unmaintainable levels and in places where it's hard to predict (more _which_ switch will grow vs "you add another case"). You add twelve types, now this switch needs the twelve types added, the file gets too long so you split the file in two, and now you dont have the  co-locality of the code enforcing the updating of the switch statement, things silently start to fall into the default case  because new engineers dont know to check this other file, now it's a whole mess. fictional but representative. so where applicable, we've been trying to move these interfaces into the type definition at the point of definition. in this case it would be moving these switch cases into `validate` functions within each of the COMMON_THRESHOLD_TYPES and having this function just call that validate function based on the type of the threshold being passed in.

Reply:
If we have the validate statements in the object COMMON_THRESHOLD_TYPES, then we have to maintain each threshold type (9 types) separately. Instead of a single switch statement with only 3 branches (grouping all numerics, versus all non-numerics, versus the deadman).

As for the number of cases possible -- that is limited by the number of possible operators (e.g. >=. =<) which are far fewer than in a redux switch statement.

But if we want to make sure that type-checking confirms that all cases are handled (a.k.a. exhaustive switch), then we can do that with use of the `never` default case. I just pushed a commit showing this ability: https://github.com/influxdata/ui/pull/4067/commits/5e6f6d975d490168c3c5a91a14b02b0c8e18815d

Reply:
Your approach (to keep everything in that object) would make sense if we had dynamic operators, and are loading an object of dynamic operators that were defined at runtime. Or we were passing around dynamically defined closures.

But instead we have a set of statically defined operators, without the ability to have field schema checking prior to query runtime (hitting the storage layer). We do have the ability, however, to pull an implicit schema (as per the MetricsSelector). The next PR will be pulling this schema per measurement. At this point, we can have the validateThreshold() method handle the enum/operator set (which is statically defined) in comparison to this implicit schema.

Using the switch statement approach allows us to quickly mix/match (in the very near future) the different operator sets, with the schema type. **This will give us UI predicate validation** on user form input...which we don't have currently. And do so in a concise method.

Writing this logic within the COMMON_THRESHOLD_TYPES object would translate to 9 different points of potential drift (versus a single function for validation control flow). And our single point of validation can turn on/off abilities more quickly (e.g. deciding if `>` operator is a numeric only, or also non-numeric).